detox/local-cli/build-framework-cache.js

```diff
@@ -1,16 +1,95 @@
-const cp = require('child_process');
 const os = require('os');
 const path = require('path');
 
+const exec = require('child-process-promise').exec;
+const fs = require('fs-extra');
+
 const detox = require('../internals');
 
 module.exports.command = 'build-framework-cache';
-module.exports.desc = 'Builds a cached Detox framework for the current environment in ~/Library/Detox. The cached framework is unique for each combination of Xcode and Detox version. (macOS only)';
+module.exports.desc = 'Builds or rebuilds a cached Detox framework and/or XCUITest-runner for the current ' +
+  'environment in ~/Library/Detox. Supports --build and --clean options to specify what to build or ' +
+  'clean. Defaults to building all. (MacOS only)';
+
+module.exports.builder = (yargs) => {
+  return yargs
+    .option('build', {
+      describe: 'Specifies what to build (if any): "all", "detox", "xcuitest", or "none"',
+      type: 'string',
+      default: 'all',
+      choices: ['all', 'detox', 'xcuitest', 'none']
+    })
+    .option('clean', {
+      describe: 'Specifies what to clean before building (if any): "all", "detox", "xcuitest", or "none"',
+      type: 'string',
+      default: 'all',
+      choices: ['all', 'detox', 'xcuitest', 'none']
+    });
+};
 
-module.exports.handler = async function buildFrameworkCache() {
-  if (os.platform() === 'darwin') {
-    cp.execSync(path.join(__dirname, '../scripts/build_framework.ios.sh'), { stdio: 'inherit' });
-  } else {
+module.exports.handler = async function buildFrameworkCache(argv) {
+  if (!isMacOS()) {
     detox.log.info(`The command is supported only on macOS, skipping the execution.`);
+    return;
   }
+
+  const frameworkPath = path.join(os.homedir(), '/Library/Detox/ios/framework');
+  const xcuitestPath = path.join(os.homedir(), '/Library/Detox/ios/xcuitest-runner');
+
+  await handleCleaning(
```

Comment:
Uh-oh... 😢  

Actually, what I meant was:

* to call `handleBuilding` in `build-framework-cache`
* to call `handleCleaning` in `clean-framework-cache`
* to call both `handleCleaning` and `handleBuilding` in `rebuild-framework-cache`

I envisioned that those utility functions will reside in separate files, and be imported by each CLI tool independently, instead of dirty tricks with using `yargs` while you are using `yargs`:

```js
  const argv = hideBin(process.argv);
  yargs(argv)
  .command(require('./build-framework-cache'))
  .parse(['build-framework-cache', '--clean="all"', '--build="all"']);
  ```

In such case, the implementation should have been dead simple:

```
handleCleaning({ xcui: boolean; framework: boolean; }) => Promise<void>
handleBuilding({ xcui: boolean; framework: boolean; }) => Promise<void>
```

Sorry for nitpicking, but wouldn't `detox (re|build|clean)-framework-cache --this --that --those` look and feel more natural and simple?

Reply:
I mean that this API doesn't make sense to me:

> --build: Specifies binaries to build (all / detox / xcuitest / none). Defaults to all.
> --clean: Specifies binaries to clean (all / detox / xcuitest / none). Defaults to none.

What sense does it bear to call `detox ... --build xcuitest --clean detox` ?

Reply:
Here's a more polite and grammatically corrected version of your message:

For `detox build-framework-cache --clean --xcuitest --detox`, the default setting should enable both `xcuitest` and `detox` enabled. However, if only one of these options is specified, the other should be disabled. This setup may not be very intuitive, but I'm okay with it. I'd prefer not to spend too much time on this internal utility to avoid unnecessary bike-shedding. I'm fine with that, will change

Reply:
For `detox build-framework-cache --clean --xcuitest --detox`, the default setting should enable both `xcuitest` and `detox` enabled. However, if only one of these options is specified, the other should be disabled. This setup may not be very intuitive, but I'm okay with it. I'd prefer not to spend too much time on discussing this internal utility to avoid bike-shedding. I'm totally fine with that, will change

Reply:
Actually, something bothers me. What's the point of adding the `--clean` flag here if we have the `rebuild-...` command? It's like using `--build` from `clean-...`
I'll just add the `--xcuitest` and `--detox` flags for all commands, and the `clean` / `rebuild` command will call the clean method.